src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs

```diff
@@ -884,56 +884,34 @@ private void DecodeModuleInitializerAttribute(DecodeWellKnownAttributeArguments<
         private void DecodeUnmanagedCallersOnlyAttribute(ref DecodeWellKnownAttributeArguments<AttributeSyntax, CSharpAttributeData, AttributeLocation> arguments)
         {
             Debug.Assert(arguments.AttributeSyntaxOpt != null);
-            if (!IsStatic || MethodKind is not (MethodKind.Ordinary or MethodKind.LocalFunction))
-            {
-                // `UnmanagedCallersOnly` can only be applied to ordinary static methods or local functions.
-                arguments.Diagnostics.Add(ErrorCode.ERR_UnmanagedCallersOnlyRequiresStatic, arguments.AttributeSyntaxOpt.Location);
-            }
 
             arguments.GetOrCreateData<MethodWellKnownAttributeData>().UnmanagedCallersOnlyAttributeData =
                 DecodeUnmanagedCallersOnlyAttributeData(arguments.Attribute, arguments.AttributeSyntaxOpt.Location, arguments.Diagnostics);
 
-            var (returnTypeSyntax, genericTypeParameters) = SyntaxNode switch
+            bool reportedError = CheckAndReportValidUnmanagedCallersOnlyTarget(arguments.AttributeSyntaxOpt.Name.Location, arguments.Diagnostics);
+
+            var returnTypeSyntax = SyntaxNode switch
```

Comment:
>SyntaxNode switch [](start = 35, length = 17)

It looks like we have an existing helper for that, ```ExtractReturnTypeSyntax```. Let's not invent new ways to do the same thing.